simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp

```diff
@@ -819,47 +820,45 @@ auto HdMapUtils::getPreviousLanelets(const lanelet::Id lanelet_id, const double
 
 auto HdMapUtils::isInRoute(const lanelet::Id lanelet_id, const lanelet::Ids & route) const -> bool
 {
-  return std::find_if(route.begin(), route.end(), [lanelet_id](const auto id) {
-           return lanelet_id == id;
-         }) != route.end();
+  return std::find(route.cbegin(), route.cend(), lanelet_id) != route.cend();
 }
 
 auto HdMapUtils::getFollowingLanelets(
-  const lanelet::Id lanelet_id, const lanelet::Ids & candidate_lanelet_ids, const double distance,
+  const lanelet::Id lanelet_id, const lanelet::Ids & candidate_route, const double distance,
   const bool include_self) const -> lanelet::Ids
 {
-  if (candidate_lanelet_ids.empty()) {
-    return {};
-  }
   lanelet::Ids ids;
-  double total_distance = 0.0;
-  bool found = false;
-  for (const auto id : candidate_lanelet_ids) {
-    if (found) {
-      ids.emplace_back(id);
-      total_distance = total_distance + getLaneletLength(id);
-      if (total_distance > distance) {
-        return ids;
-      }
-    }
-    if (id == lanelet_id) {
-      found = true;
-      if (include_self) {
-        ids.push_back(id);
-      }
-    }
+
+  if (candidate_route.empty()) {
+    return ids;
   }
-  if (!found) {
+
+  auto lanelet_it = std::find(candidate_route.cbegin(), candidate_route.cend(), lanelet_id);
```

Comment:
dirty == untested, take a look at this and let me know what you think
```
auto HdMapUtils::getFollowingLanelets(
  const lanelet::Id lanelet_id, const lanelet::Ids & candidate_lanelet_ids, const double distance,
  const bool include_self) const -> lanelet::Ids
{
  if (candidate_lanelet_ids.empty()) {
    return {};
  } else {
    lanelet::Ids following_lanelets_ids;
    double total_distance = 0.0;
    bool found_reference_lanelet_id = false;
    for (const auto & candidate_lanelet_id : candidate_lanelet_ids) {
      if (found_reference_lanelet_id) {
        following_lanelets_ids.push_back(candidate_lanelet_id);
        total_distance += lanelet_map::laneletLength(candidate_lanelet_id);
        if (total_distance > distance) {
          return following_lanelets_ids;
        }
      }
      if (!found_reference_lanelet_id && candidate_lanelet_id == lanelet_id) {
        found_reference_lanelet_id = true;
        if (include_self) {
          following_lanelets_ids.push_back(candidate_lanelet_id);
        }
      }
    }

    if (!found_reference_lanelet_id) {
      THROW_SEMANTIC_ERROR("lanelet id does not match");
    } else if (total_distance > distance) {
      return following_lanelets_ids;
    } else {
      const auto remaining_lanelets =
        followingLanelets(candidate_lanelet_ids.back(), distance - total_distance, false);
      following_lanelets_ids.insert(
        following_lanelets_ids.end(), remaining_lanelets.begin(), remaining_lanelets.end());
      return following_lanelets_ids;
    }
  }
}
```

Reply:
this doesn't resolve described issue, as I understand it's preferred not to work on iterators, I've changed this to check if proposed route is actually possible one.